File: ios/CreatePrivateDealerOS/CreatePrivateDealerOS/ViewModels/AnalyticsViewModel.swift
import Foundation

@MainActor
final class AnalyticsViewModel: ObservableObject {
    @Published var snapshot = AnalyticsSnapshot(
        responseTrend7d: [],
        responseTrend30d: [],
        sourceBreakdown: [],
        bookingsCount: 0,
        afterHoursHandledPercent: 0
    )
    @Published var selectedRangeDays = 7
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let repository: DealerOSRepository

    init(repository: DealerOSRepository) {
        self.repository = repository
    }

    func load(session: AppSession) async {
        guard !isLoading else { return }
        isLoading = true
        defer { isLoading = false }

        do {
            snapshot = try await repository.fetchAnalytics(session: session)
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}

File: flutter/create_private_dealer_os/lib/application/view_models/analytics_view_model.dart
import 'package:flutter/foundation.dart';

import '../../domain/models/analytics_snapshot.dart';
import '../../domain/models/app_session.dart';
import '../../domain/repositories/dealer_os_repository.dart';

class AnalyticsViewModel extends ChangeNotifier {
  AnalyticsViewModel({required DealerOsRepository repository})
      : _repository = repository,
        _snapshot = const AnalyticsSnapshot(
          responseTrend7d: <ResponseTrendPoint>[],
          responseTrend30d: <ResponseTrendPoint>[],
          sourceBreakdown: <SourceBreakdown>[],
          bookingsCount: 0,
          afterHoursHandledPercent: 0,
        );

  final DealerOsRepository _repository;

  AnalyticsSnapshot _snapshot;
  int _selectedRangeDays = 7;
  bool _isLoading = false;
  String? _errorMessage;

  AnalyticsSnapshot get snapshot => _snapshot;
  int get selectedRangeDays => _selectedRangeDays;
  bool get isLoading => _isLoading;
  String? get errorMessage => _errorMessage;

  set selectedRangeDays(int value) {
    _selectedRangeDays = value;
    notifyListeners();
  }

  Future<void> load({required AppSession session}) async {
    if (_isLoading) {
      return;
    }

    _isLoading = true;
    _errorMessage = null;
    notifyListeners();

    try {
      _snapshot = await _repository.fetchAnalytics(session: session);
    } catch (error) {
      _errorMessage = error.toString();
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }
}
